### PR TITLE
Attachment Event

### PR DESCRIPTION
**Pull request by electrostat**
Create an attachment that contains relevant metadata and binary data for video or images.

closes https://github.com/mapbox/mapbox-events-android/issues/200

### CODE REVIEW

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/Attachment.java [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+
+public class Attachment extends Event implements Parcelable {
+  private static final String VIS_ATTACHMENT = "vis.attachment";
+
+  @SerializedName("event")
+  private final String event;
+  @SerializedName("attachmentMetadata")
+  private List<AttachmentMetadata> attachmentMetadata;
+  @SerializedName("files")
+  private List<byte[]> files;
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/Attachment.java]: What are some benefits of maintaining collection of binary blobs as opposed to collection of paths to files on the filesystem? Are there any memory implications?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/Attachment.java [hunk 2]
```diff
@@ -0,0 +1,75 @@
...
+import java.util.ArrayList;
+import java.util.List;
+
+public class Attachment extends Event implements Parcelable {
+  private static final String VIS_ATTACHMENT = "vis.attachment";
+
+  @SerializedName("event")
+  private final String event;
+  @SerializedName("attachmentMetadata")
+  private List<AttachmentMetadata> attachmentMetadata;
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/Attachment.java]: Shouldn't lists be immutable?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/Attachment.java [hunk 3]
```diff
@@ -0,0 +1,75 @@
...
+  @SerializedName("files")
+  private List<byte[]> files;
+
+  Attachment() {
+    this.event = VIS_ATTACHMENT;
+    attachmentMetadata = new ArrayList<>();
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/Attachment.java]: For consistency sake, would it make sense to use "this.* =" pattern across all constructor initializations?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/Attachment.java (hunk 1 above)
Comment (electrostat) [reply libtelemetry/src/main/java/com/mapbox/android/telemetry/Attachment.java]: No memory implications. Told that we will be receiving the data as binary from host sdk. We are working with the limitations with how the vision sdk currently works.

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/VisionAttachment.java [hunk 4]
```diff
@@ -0,0 +1,23 @@
+package com.mapbox.android.telemetry;
+
+import okhttp3.MediaType;
+
+public class VisionAttachment {
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/VisionAttachment.java]: Should name of this class be more generic? One suggestion could be: FileAttachment. Thoughts?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/VisionEventFactory.java [hunk 5]
```diff
@@ -30,11 +32,32 @@ public Event createVisionEvent(Event.Type type) {
...
+    checkVisionEvent(type);
+    return buildAttachment();
+  }
+
   private VisionEvent buildVisionEvent() {
     VisionEvent visionEvent = new VisionEvent();
     return visionEvent;
   }
 
+  private VisionAttachment buildVisionAttachment(String filepath, MediaType mediaType,
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/VisionEventFactory.java]: What is the benefit of maintaining a set of private wrapper methods that simply return the instance of an object?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/VisionEventFactory.java [hunk 6]
```diff
@@ -30,11 +32,32 @@ public Event createVisionEvent(Event.Type type) {
     return BUILD_EVENT_VISION.get(type).build();
   }
 
+  public VisionAttachment createVisionAttachment(String filePath, MediaType mediaType,
+                                                  AttachmentMetadata attachmentMetadata) {
+    return buildVisionAttachment(filePath, mediaType, attachmentMetadata);
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/VisionEventFactory.java]: wouldn't this be a little more readable and less redundant if we do:
`return new VisionAttachment(attachmentMetadata, filepath, mediaType);`

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/VisionEventFactory.java [hunk 7]
```diff
@@ -30,11 +32,32 @@ public Event createVisionEvent(Event.Type type) {
     return BUILD_EVENT_VISION.get(type).build();
   }
 
+  public VisionAttachment createVisionAttachment(String filePath, MediaType mediaType,
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/VisionEventFactory.java]: Shouldn't we start annotating input variables of all public methods?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryClient.java [hunk 8]
```diff
@@ -50,6 +53,40 @@ void sendEvents(List<Event> events, Callback callback) {
     sendBatch(batch, callback);
   }
 
+  void sendAttachment(Attachment attachment, Callback callback) {
+    List<VisionAttachment> visionAttachments = attachment.getAttachments();
+    List<AttachmentMetadata> metadataList = new ArrayList<>();
+
+    MultipartBody.Builder requestBodyBuilder = new MultipartBody.Builder("--01ead4a5-7a67-4703-ad02-589886e00923")
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryClient.java]: I wonder why this value is hardcoded and if there are any downsides to hardcoding it?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryClient.java [hunk 9]
```diff
@@ -50,6 +53,40 @@ void sendEvents(List<Event> events, Callback callback) {
     sendBatch(batch, callback);
   }
 
+  void sendAttachment(Attachment attachment, Callback callback) {
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryClient.java]: What are some ways we can improve readability of this method? Can we split the content of it into a few helper methods?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryClientSettings.java [hunk 10]
```diff
@@ -136,13 +140,17 @@ TelemetryClientSettings build() {
     }
   }
 
-  private OkHttpClient configureHttpClient() {
+  private OkHttpClient configureHttpClient(boolean attachment) {
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/TelemetryClientSettings.java]: What are some benefits and side-effects of passing a boolean flag as opposed to passing a ref to interceptor itself? One specific suggestion here is to turn this signature to:
`private OkHttpClient configureHttpClient(@Nullable Interceptor inteceptor){}`
and check if inteceptor ref is passed to this method?

then your client method might look like: 
```
OkHttpClient getClient() {
   return configureHttpClient(new GzipRequestInterceptor());
}
```
Thoughts?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/Attachment.java [hunk 11]
```diff
@@ -0,0 +1,62 @@
...
+
+import android.os.Parcel;
+import android.os.Parcelable;
+
+import com.google.gson.annotations.SerializedName;
+
+import java.util.ArrayList;
+import java.util.List;
+
+public class Attachment extends Event implements Parcelable {
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/Attachment.java]: I wonder if we could get to a more consistent naming conventions here? Attachment vs. VisionAttachment: should attachment event be more generic?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/AttachmentFilePath.java [hunk 12]
```diff
@@ -0,0 +1,13 @@
+package com.mapbox.android.telemetry;
+
+import okhttp3.MediaType;
+
+class AttachmentFilePath {
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/AttachmentFilePath.java]: Can we think of a better name for this class? I know names are tough...

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/AttachmentFilePath.java [hunk 13]
```diff
@@ -0,0 +1,13 @@
...
+  String filePath;
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/AttachmentFilePath.java]: Should these fields be immutable by default?

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/AttachmentMetadata.java [hunk 14]
```diff
@@ -0,0 +1,65 @@
...
+  private String name;
+  private String created;
+  private String eventId;
+  private String format;
+  private String type;
+  private Integer size;
+  private String startTime;
+  private String endTime;
+
+  public AttachmentMetadata(String name, String eventId, String format, String type) {
```
Comment (andrlee) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/AttachmentMetadata.java]: bracket on newline

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/AttachmentMetadata.java [hunk 15]
```diff
@@ -0,0 +1,71 @@
+package com.mapbox.android.telemetry;
+
+public class AttachmentMetadata {
+
+  private String name;
+  private String created;
+  private String fileId;
+  private String format;
+  private String type;
+  private String sessionId;
```
Comment (BenDenen) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/AttachmentMetadata.java]: Is it ok for server side that we use String as ID - if yes I am ok with this

Comment (electrostat) [reply libtelemetry/src/main/java/com/mapbox/android/telemetry/AttachmentMetadata.java]: Confirmed that we are looking for strings server side for the IDs.

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/AttachmentListener.java [hunk 16]
```diff
@@ -0,0 +1,9 @@
+package com.mapbox.android.telemetry;
+
+import java.util.List;
+
+public interface AttachmentListener {
+  void onAttachmentResponse(String message, int code, List<String> fieldIds);
```
Comment (yunikkk) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/AttachmentListener.java]: Seems that here is a typo and `fieldIds` should be `fileIds` instead

Comment (electrostat) [reply libtelemetry/src/main/java/com/mapbox/android/telemetry/AttachmentListener.java]: Being handled in this PR: https://github.com/mapbox/mapbox-events-android/pull/212

File: libtelemetry/src/main/java/com/mapbox/android/telemetry/MapboxTelemetry.java [hunk 17]
```diff
@@ -460,6 +475,11 @@ private boolean sendEventIfWhitelisted(Event event) {
       sendEventsIfPossible(appUserTurnstile);
       return true;
     }
+
+    if (Event.Type.VIS_ATTACHMENT.equals((event.obtainType()))) {
+      sendAttachment(event);
```
Comment (yunikkk) [inline libtelemetry/src/main/java/com/mapbox/android/telemetry/MapboxTelemetry.java]: Should we return `true` here as the event is already sent?
#225

**Pull request merged**